### PR TITLE
Reintroduce updated backendutil.TransformBackend after 1a1363a

### PR DESCRIPTION
**Pull request by mback2k**
After considering the comments in https://github.com/emersion/go-smtp/issues/22 I still think that a TransformBackend could be useful. Especially for me since it is required for my little program https://github.com/mback2k/smtp-dkim-signer.

Considering the comments in https://github.com/emersion/go-smtp/issues/22#issuecomment-475669004 I think that option 2) "Split Transform function into separate TransformFrom, TransformTo, TransformData. Pass transformed data to underlying backend instantly." makes totally sense. The mentioned aspects are:
- The underlying backend does have an ability to report errors on From/Rcpt => true
- Can't use From/To information in TransformData =>  still possible since the required state can be kept with a Session-based approach

Therefore I am proposing to bring back TransformBackend with a Session-based approach of individual Transform functions that could be bound to an object themselves. I will post an example in a separate comment. What do you think?

### CODE REVIEW

**Comment by mback2k** [comment]
Example usage can be seen here: https://github.com/mback2k/smtp-dkim-signer/blob/674e0720641237d6166dc903e37aedc180e1a495/backend.go#L58-L151

File: backendutil/transform.go [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+// TransformBackend is a backend that transforms messages.
+type TransformBackend struct {
+	Backend smtp.Backend
+
+	Transform          func(session *smtp.Session, username string) TransformHandler
+	AnonymousTransform func(session *smtp.Session) TransformHandler
+}
+
+// TransformHandler is a container for transforming funcs.
+type TransformHandler interface {
```
Comment (foxcpp) [inline backendutil/transform.go]: This requires users to create an object that implements TransformHandler, right? Isn't it the same as implementing the actual Session interface? I guess you wanted to have a structure with func(..) variables here.

Comment (mback2k) [reply backendutil/transform.go]: Thanks. You are right, I probably wanted a struct instead of an interface. Will make the change later today.

**Comment by mback2k** [comment]
@emersion @foxcpp I am also thinking about whether the TransformBackend should be part of this module in the package backendutil or rather a separate module, just like go-smtp-proxy. What do you think? That way it could be patched / updated / changed independently.

**Comment by foxcpp** [comment]
It is relatively small (106 loc, by looking at your commits) so I don't think it should be maintained separately, it could create too much micromanagement with versions and stuff. Additionally, I think TransformBackend could often be used with simple proxy servers that do one thing (like your DKIM signer), so users could benefit from having it together with go-smtp.

**Comment by mback2k** [comment]
Sounds good. The current proposal basically implements pre-hooks for the Session methods. I think it would make sense to have post-hooks to be able to handle errors as well. What do you think?

**Comment by foxcpp** [comment]
Post-hooks are required so you can rollback the internal state of "transformer" when the command gets rejected by go-smtp for whatever reason, right? TBH I actually doubt the idea of having state hidden behind functions. If you need some per-session state, you probably should just implement the Session interface and use it, this will make your intent more clear. I think TransformBackend functions should be pure functions, after all, name indicates what they transform, tranforming something else would be... ehh.

**Comment by mback2k** [comment]
> Post-hooks are required so you can rollback the internal state of "transformer" when the command gets rejected by go-smtp for whatever reason, right?

Yes, that is the purpose I was thinking about.

> TBH I actually doubt the idea of having state hidden behind functions. If you need some per-session state, you probably should just implement the Session interface and use it, this will make your intent more clear. I think TransformBackend functions should be pure functions, after all, name indicates what they transform, tranforming something else would be... ehh.

You are probably right. That means the TransformBackend is useful in itself for transforming purposes, but for my specific use case I will just implement the Session interface.

Do we still want to have Session-based Transform functions then? I guess it could still make sense to be able to do different transformations depending on the session / username. Otherwise we could just have the 4 (optional and nil-able) transform functions on the Backend-level again.

**Comment by foxcpp** [comment]
>Otherwise we could just have the 4 (optional and nil-able) transform functions on the Backend-level again.

Yeah, I guess that's how it should be done.

**Comment by emersion** [comment]
Agree with @foxcpp

**Comment by mback2k** [comment]
@foxcpp @emersion Thanks for the feedback. I just incorporated all of it and pushed a new commit.

I guess for a final PR all of the individual commits can be squashed together before merge.

If you are fine with the new state, I will do that and also try to add some tests.

**Comment by foxcpp** [comment]
Feel free to add more commits, we can always squash&merge it.

File: backendutil/transform.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+	"io"
+
+	"github.com/emersion/go-smtp"
+)
+
+// TransformBackend is a backend that transforms messages.
+type TransformBackend struct {
+	Backend smtp.Backend
+
+	TransformMail func(from string) (string, error)
```
Comment (foxcpp) [inline backendutil/transform.go]: It would be nice to add a usage example for it, wdyt?

Comment (mback2k) [reply backendutil/transform.go]: I just pushed some tests for the transform backend based on the existing server tests.

Since I am not using the TransformBackend anymore, I was not able to think of a good example. For the tests I just made the transform functions encode everything using Base64.

**Comment by mback2k** [comment]
I think this PR is now ready to be merged. I am looking forward to your final feedback @emersion @foxcpp or for the merge. :-)

File: server_test.go [hunk 3]
```diff
@@ -405,7 +405,7 @@ func TestServer_tooManyInvalidCommands(t *testing.T) {
 }
 
 func TestServer_tooLongMessage(t *testing.T) {
```
Comment (foxcpp) [inline server_test.go]: Is this test relevant to TransformBackend functionality?

File: backendutil/transform_test.go [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+	if len(msg.To) != 1 || msg.To[0] != "cm9vdEBnY2hxLmdvdi51aw==" {
+		t.Fatal("Invalid mail recipients:", msg.To)
+	}
+	// base64 of "Hey <3\n" (with actual newline)
+	if string(msg.Data) != "SGV5IDwzCg==" {
+		t.Fatal("Invalid mail data:", string(msg.Data))
+	}
+}
+
+func TestServer_tooLongMessage(t *testing.T) {
```
Comment (foxcpp) [inline backendutil/transform_test.go]: Is this test relevant to TransformBackend functionality?

Comment (mback2k) [reply backendutil/transform_test.go]: Yes, it is there to check if errors are passed through correctly. I missed that part without that test initially. ;-)

Comment (foxcpp) [reply backendutil/transform_test.go]: Alright.

**Comment by emersion** [comment]
Thanks!
